WooCommerce/Classes/POS/ViewModels/ItemSelectorViewModel.swift

```diff
@@ -4,14 +4,30 @@ import protocol Yosemite.POSItem
 import protocol Yosemite.POSItemProvider
 
 final class ItemSelectorViewModel: ObservableObject {
-    let selectedItemPublisher: AnyPublisher<POSItem, Never>
+    enum ItemSelectorState {
```

Comment:
As I understand, the scope of this task is only to provide a skeleton for the eventual proper handling of `loading`, `loaded`, and `error` states. However, we may want to set the structure for the future tasks as well:

Usually, when the view is controlled by a state enum, it would also contain the associated values. Therefore instead of having both `items`, and `state`, we would only have `state`, which would also set `items` for its loaded state. `state = .loaded(items)`. WDYT?

Reply:
That sounds good, my initial implementation used associated values for `.loaded(hasItems: Bool?)` since we will have different states based on no items (nil), or eligible items (true)/non-eligible items (false). But was too many changes for the scope so I cuted it short for now. We can iterate over this this on future PRs 👍

Reply:
Yep you're right, the `nil` case doesn't look great, but I'm also not convinced yet of marking `.nonEligibleItems` as an error 🤔 I'm thinking something like the following would make sense, since at the end of the day both no items and non eligible are `.loaded([])`, just for different reasons:

```
    enum ItemListState<T> {
        case loading
        case loaded(items: [T], reason: Reason)
        case error(Error)
        
        enum Reason {
            case noItems
            case noEligibleItems
        }
    }
```
I'll play a bit with it and open a discussion to get some input and parity between platforms! 👍

Reply:
Good, thanks!

>  case loaded(items: [T], reason: Reason)

In this case, the view could receive a proper list of items, but Reason: `.noEligibleItems` or `.noItems`. Which in reality shouldn't exist but the code paths would need to be handled. 

In the designs, we do distinguish between an error state and no product state. It could imply that the easiest path would be to have a different type of state for them: 
<img width="809" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/4062343/ebb6946c-6514-4676-ba9e-f599029a8ffb">
<img width="814" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/4062343/b97400f6-a0eb-4645-bf53-9c0c89cda555">

Then our `View` wouldn't need to make any decisions and only blindly present:

```
switch state {
   case .loaded(let items):
       ListView(items)
    case .error(let errorVM):
       ErrorView(errorVM)
    case .empty(let emptyVM):
        EmptyView(emptyVM)
    case .loading:
        LoadingView()
}
```

In other cases, the view would contain logic such as `if items.empty && reason == .noEligibleItems` to decide what to present.